### ISSUE TITLE
Do Expressions

### ISSUE THREAD

**Issue by hzoo** [estree/estree#155]
Babel docs: http://babeljs.io/docs/plugins/transform-do-expressions
Stage 1 gist reproduced below: https://gist.github.com/dherman/1c97dfb25179fa34a41b5fff040f9879 by @dherman 

In Babel

```js
interface DoExpression <: Expression {
  type: "Identifier";
  body: BlockStatement
}
```

---

# Motivation

* expression-oriented programming one of the great advances of FP
* expressions plug together like legos, making more malleable programming experience in-the-small

# Examples

Write in an expression-oriented style, scoping variables as locally as possible:

```js
let x = do {
  let tmp = f();
  tmp * tmp + 1
};
```

Use conditional statements as expressions, instead of awkward nested ternaries:

```js
let x = do {
  if (foo()) { f() }
  else if (bar()) { g() }
  else { h() }
};
```

Especially nice for templating languages like JSX:

```js
return (
  <nav>
    <Home />
    {
      do {
        if (loggedIn) {
          <LogoutButton />
        } else {
          <LoginButton />
        }
      }
    }
  </nav>
)
```
```

**Comment by mikesherov** [comment]
Seems fairly straightforward to me. @dherman any footguns to be aware of? @rreverser and @ariya any thoughts?

**Comment by ariya** [comment]
I haven't investigated do expressions yet.

BTW, why is the type `Identifier`?

**Comment by michaelficarra** [comment]
One of the major controversies with the proposal within the committee is whether abrupt completions such as return/break/continue statements should be allowed within do expressions.

**Comment by RReverser** [comment]
> BTW, why is the type Identifier?

Seems like a copy-paste mistake to me.

**Comment by hzoo** [comment]
Oops, fixed (good catch @ariya)

**Comment by gibson042** [comment]
> `type: "BlockStatement"`

Don't you mean `type: "DoExpression"`?

**Comment by mikesherov** [comment]
@gibson042, isn't the body of a do Expression just a normal block for the case of tree representation?

**Comment by gibson042** [comment]
Yes, subject to the concerns about completions @michaelficarra mentioned. But `{ "type: "BlockStatement", "body": […] }` is an existing AST node, and [it isn't an Expression](https://github.com/estree/estree/blob/master/es5.md#blockstatement).

**Comment by mikesherov** [comment]
The body is a blockStatement (subject to abrupt completions), the type is "DoExpression"... no? what am I missing?

**Comment by RReverser** [comment]
@mikesherov I think when @gibson042 wrote his previous comment

> Don't you mean type: "DoExpression"?

main type of node was "BlockExpression", but @hzoo edited it afterwards, so it's now "DoExpression", thus misunderstanding :)
